### ISSUE TITLE
Improvement: parallelize `AutoML.predict`

### ISSUE THREAD

**Issue by lsorber** [automl/auto-sklearn#239]
Prediction with an ensemble can take a long time depending on the number of models. A quick win would be to use a `multiprocessing.Pool.map` to replace this for loop in `AutoML.predict` (line 455-470):

```
all_predictions = []
for identifier in self.ensemble_.get_model_identifiers():
    model = self.models_[identifier]

    X_ = X.copy()
    if self._task in REGRESSION_TASKS:
        prediction = model.predict(X_)
    else:
        prediction = model.predict_proba(X_)

    if len(prediction.shape) < 1 or len(X_.shape) < 1 or \
            X_.shape[0] < 1 or prediction.shape[0] != X_.shape[0]:
        self._logger.warning("Prediction shape for model %s is %s "
                             "while X_.shape is %s" %
                             (model, str(prediction.shape),
                              str(X_.shape)))
    all_predictions.append(prediction)
```

**Comment by mfeurer** [comment]
Indeed, this would help, although I think one should use `sklearn.externals.joblib`. A PR would be highly appreciated.

This is related to #75.

**Comment by lsorber** [comment]
Unfortunately, I don't have a lot of time on my hands right now, but I might give it a go soon-ish. If I do, I will look into using `joblib` as you suggest.

Another interesting note is that some models (I'm looking at you Naive Bayes) run out of memory predicting on a large number of rows (e.g., `X.shape[0] > 100e3`). Hence, chunking the predictions into batches of 20k rows seams like a good idea to incorporate into this PR.

**Comment by mfeurer** [comment]
The underlying [predict function](https://github.com/automl/auto-sklearn/blob/master/autosklearn/pipeline/base.py#L155) support chunking and is already used while fitting the models at optimization time. I have never experienced this at prediction time though, that's why it's not active.

**Comment by lsorber** [comment]
All right, I'll see if I can free up some time over the weekend for a PR that uses `joblib` and sets the `batch_size`.

**Comment by mfeurer** [comment]
That'd be great. I'll be on holiday from next week, so I won't be able to respond and merge this for a bit.

**Comment by Motorrat** [comment]
you could use this function to determine the number of available processors 
```
import psutil, math
memory_limit = 15000 # MB
def define_pool_size(memory_limit):
    # some classifiers can use more than one core - so keep this at half memory and cores
    max_pool_size = int(math.ceil(psutil.virtual_memory().total / (memory_limit * 1000000)))
    half_of_cores = int(math.ceil(psutil.cpu_count()/2.0))
    half_of_cores if max_pool_size > half_of_cores else max_pool_size
    return half_of_cores if max_pool_size > half_of_cores else max_pool_size
```

**Comment by lsorber** [comment]
@mfeurer The PR should be ready, let me know if you need any changes. Out of curiosity, is there any particular reason you prefer `joblib` over `multiprocessing`?

@Motorrat Thanks for the suggestion. In the PR, I set `n_jobs=1` so that users have the option to predict in parallel, while keeping the default behaviour the same as before.

**Comment by mlindauer** [comment]
@lsorber sorry, @mfeurer is on vacation for 3 weeks and the PR has to wait until he is back.

**Comment by lsorber** [comment]
No problem, just ping me if you need me.

**Comment by Motorrat** [comment]
@lsorber my point is that n_jobs=-1 doesn't necessarily work "out of the box". So a proper solution should consider the limitations of the actual hardware and the dataset.

**Comment by mfeurer** [comment]
Fixed by #242
